src/main/scala/typechecker/Subtyping.scala

```diff
@@ -57,21 +57,33 @@ object Subtyping {
     case (TSizedInt(v1, un1), TSizedInt(v2, un2)) => un1 == un2 && v1 <= v2
     case (TStaticInt(v1), TSizedInt(v2,un2)) => ( (v1<0 && un2==false) | (v1>=0) ) && bitsNeeded(v1)<=v2 
     case (_:TIndex,_:TSizedInt) => true   
-    //case (_:IntType, _:TSizedInt) => true
     case (_:TStaticInt, _:TIndex) => true
     case (TArray(tsub, subDims), TArray(tsup, supDims)) => {
       // Arrays are invariant
       areEqual(tsup, tsub) && subDims == supDims
     }
     case (_:TFloat, _:TDouble) => true
+    case (_:TRational, _:TFloat) => true
+    case (_:TRational, _:TDouble) => true
+    case (TRational(v1), TFixed(_,i2,un2)) => 
+      ( (v1.toDouble<0 && un2==false) | (v1.toDouble>=0) ) && bitsNeeded(v1.toDouble.toInt)<=i2 
+    case (TFixed(t1,i1,un1), TFixed(t2,i2,un2) ) => (un1 == un2 && i1 <= i2 && (t1-i1)<=(t2-i2) )
     case _ => areEqual(sub, sup)
   }
 
   private def joinOfHelper(t1: Type, t2: Type, op: BOp): Option[Type] = (t1, t2) match {
+    //XXX(Zhijing): what happens for multiplication? Overflow?
     case (TStaticInt(v1), TStaticInt(v2)) => op.toFun match {
       case Some(fun) => Some(TStaticInt(fun(v1, v2)))
       case None => Some(TSizedInt(max(bitsNeeded(v1), bitsNeeded(v2)), false))
     }
+    case (TRational(v1), TRational(v2)) => op.toFun match {
+      //XXX(Zhijing):deprecated
+      case Some(fun) => Some(TRational(fun(v1.toDouble.toInt, v2.toDouble.toInt).toString))
```

Comment:
This seems wrong. op.toFun returns a function that does type level computation for naturals.

Here, you’re doing a computation by converting the rational values to integers (losing precision).

If you want to do type level computation, define a toDoubleFun function and do a precise calculation. Doing this does break the promise that rationals are infinite precision.

Reply:
There is no way (or too troublesome) to do a precise calculation because that's the natural of floating point calculation in PC. I can either remove toInt, or just no long do any computation.

Reply:
If you remove it, you have to choose a type to retune for these computations. If you return double, it’ll fail for

let x: fix<3, 3> = 1.2 + 3.4

Reply:
No. I just keep the one with larger range in it's integer part. See the None case:
if (bitsNeeded(v1.toDouble.toInt)>bitsNeeded(v2.toDouble.toInt)) Some(TRational(v1))
        else Some(TRational(v2))

Reply:
Maybe I'm not following what this code is supposed to do. Can you explain what should happen here?

The `Some` case triggers for all arithmetic operators.
The `None` case only triggers when you compare two numbers. (Look at the calls to `joinOf` in TypeCheck.scala)